crates/unftp-sbe-fs/src/lib.rs

```diff
@@ -69,48 +79,37 @@ impl Filesystem {
     /// asks for `hello.txt`, the server will send it `/srv/ftp/hello.txt`.
     pub fn new<P: Into<PathBuf>>(root: P) -> Self {
         let path = root.into();
-        Filesystem {
-            root: canonicalize(&path).unwrap_or(path),
-        }
-    }
-
-    /// Returns the full, absolute and canonical path corresponding to the (relative to FTP root)
-    /// input path, resolving symlinks and sequences like '../'.
-    async fn full_path<P: AsRef<Path>>(&self, path: P) -> Result<PathBuf> {
-        // `path.join(other_path)` replaces `path` with `other_path` if `other_path` is absolute,
-        // so we have to check for it.
-        let path = path.as_ref();
-        let full_path = if path.starts_with("/") {
-            self.root.join(path.strip_prefix("/").unwrap())
-        } else {
-            self.root.join(path)
-        };
-
-        let real_full_path = tokio::task::spawn_blocking(move || canonicalize(full_path))
-            .await
-            .map_err(|e| Error::new(ErrorKind::LocalError, e))??;
-
-        if real_full_path.starts_with(&self.root) {
-            Ok(real_full_path)
-        } else {
-            Err(Error::from(ErrorKind::PermanentFileNotAvailable))
-        }
+        let aa = cap_std::ambient_authority();
+        let root_fd = Arc::new(cap_std::fs::Dir::open_ambient_dir(&path, aa).unwrap());
+        Filesystem { root_fd, root: path }
     }
 }
 
 #[async_trait]
 impl<User: UserDetail> StorageBackend<User> for Filesystem {
     type Metadata = Meta;
 
+    /// Restrict the backend's capabilities so that it may only access files underneath `path`.
```

Comment:
```rust
    /// Restrict the backend's capabilities so that it may only access files underneath `path`.
    /// Once restricted, it may never be unrestricted.
    fn enter(&mut self, user: &User) -> io::Result<()> {
        if let Some(path) = user.home() {
            let relpath = match path.strip_prefix(self.root.as_path()) {
                Ok(r) => r,
                Err(_) => return Err(io::Error::new(io::ErrorKind::Other, "Path not a descendant of the previous root")),
            };
            self.root_fd = Arc::new(self.root_fd.open_dir(relpath)?);
        }
        Ok(())
    }
```

when we modify the `enter` method to take a `UserDetail` implementation instead of a path.